tpm2/transport/tcp/tcp.go

```diff
@@ -0,0 +1,263 @@
+	platformACTGetSignaled          platformCommand = 26
+	platformTestFailureMode         platformCommand = 30
+	platformSetFWHash               platformCommand = 35
+	platformSetFWSVN                platformCommand = 36
+)
+
+func (c platformCommand) String() string {
+	switch c {
+	case platformPowerOn:
+		return "POWER_ON"
+	case platformPowerOff:
+		return "POWER_OFF"
+	case platformPPOn:
+		return "PP_ON"
+	case platformPPOff:
+		return "PP_OFF"
+	case platformCancelOn:
+		return "CANCEL_ON"
+	case platformCancelOff:
+		return "CANCEL_OFF"
+	case platformNVOn:
+		return "NV_ON"
+	case platformNVOff:
+		return "NV_OFF"
+	case platformKeyCacheOn:
+		return "KEY_CACHE_ON"
+	case platformKeyCacheOff:
+		return "KEY_CACHE_OFF"
+	case platformReset:
+		return "RESET"
+	case platformRestart:
+		return "RESTART"
+	case platformSessionEnd:
+		return "SESSION_END"
+	case platformStop:
+		return "STOP"
+	case platformGetCommandResponseSizes:
+		return "GET_COMMAND_RESPONSE_SIZES"
+	case platformACTGetSignaled:
+		return "ACT_GET_SIGNALED"
+	case platformTestFailureMode:
+		return "TEST_FAILURE_MODE"
+	case platformSetFWHash:
+		return "SET_FW_HASH"
+	case platformSetFWSVN:
+		return "SET_FW_SVN"
+	default:
+		return fmt.Sprintf("unknown platform command (%v)", uint32(c))
+	}
+}
+
+type TPM struct {
+	cmd  *net.TCPConn
+	plat *net.TCPConn
+}
+
+type tpmCommandHeader struct {
+	tcpCommand regularCommand
+	locality   uint8
+	cmdLen     uint32
+}
+
+// Send implements the TPMCloser interface.
+func (t *TPM) Send(cmd []byte) ([]byte, error) {
+	hdr := tpmCommandHeader{
+		tcpCommand: tpmSendCommand,
+		locality:   0,
+		cmdLen:     uint32(len(cmd)),
+	}
+	// Write the header followed by the request
+	if err := binary.Write(t.cmd, binary.BigEndian, hdr); err != nil {
+		return nil, fmt.Errorf("%w: could not send TPM command to service: %v", ErrTransport, err)
+	}
+	if n, err := t.cmd.Write(cmd); err != nil {
+		return nil, fmt.Errorf("%w: could not send TPM command to service: %v", ErrTransport, err)
+	} else if n != len(cmd) {
+		return nil, fmt.Errorf("%w: could not send full TPM command: only sent %v out of %v bytes", ErrTransport, n, len(cmd))
+	}
+
+	// Read the response
+	var rspLen uint32
+	if err := binary.Read(t.cmd, binary.BigEndian, &rspLen); err != nil {
+		return nil, fmt.Errorf("%w: could not read TPM response from service: %v", ErrTransport, err)
+	}
+	if rspLen > maxBufferSize {
+		return nil, fmt.Errorf("%w: response (%v bytes) was bigger than max size (%v bytes)", ErrResponseTooBig, rspLen, maxBufferSize)
+	}
+	rsp := make([]byte, int(rspLen))
+	if n, err := t.cmd.Read(rsp); err != nil {
+		return nil, fmt.Errorf("%w: could not read TPM response from service: %v", ErrTransport, err)
+	} else if n != len(rsp) {
+		return nil, fmt.Errorf("%w: could not read full TPM response: only got %v out of %v bytes", ErrTransport, n, len(rsp))
+	}
+	// The server also provides a TCP error code at the end.
```

Comment:
Where is this in the implementation? I couldn't find it.

Reply:
That would be here: https://github.com/TrustedComputingGroup/TPM/blob/17481837510a0c6e7878455d8256cb669e755e96/TPMCmd/Simulator/src/TcpServer.c#L742